File: NotesTracker.Business/Services/NotesService.cs
// *********************************************************************************
//	<copyright file="NotesService.cs" company="Personal">
//		Copyright (c) 2025 Personal
//	</copyright>
// <summary>The Notes Service Interface.</summary>
// *********************************************************************************

namespace NotesTracker.Business.Services
{
	using NotesTracker.Business.Contracts;
	using NotesTracker.Data.Contracts;
	using NotesTracker.Data.Entities;
	using NotesTracker.Shared.Constants;
	using NotesTracker.Shared.DTO;

	/// <summary>
	/// The Notes Service Class.
	/// </summary>
	/// <seealso cref="NotesTracker.Business.Contracts.INotesService" />
	public class NotesService(INotesDataService notesDataService) : INotesService
	{
		/// <summary>
		/// The notes data service
		/// </summary>
		private readonly INotesDataService _notesDataService = notesDataService;

		/// <summary>
		/// Adds the new note asynchronous.
		/// </summary>
		/// <param name="newNote">The new note.</param>
		/// <returns>
		/// The boolean for success/failure
		/// </returns>
		/// <exception cref="System.ArgumentNullException">newNote</exception>
		public Task<bool> AddNewNoteAsync(NoteDTO newNote)
		{
			HandleUserValidation(newNote.UserName);
			if (newNote is not null)
			{
				return this._notesDataService.AddNewNoteAsync(newNote);
			}

			throw new ArgumentNullException(nameof(newNote), ExceptionConstants.NewNoteNullException);
		}

		/// <summary>
		/// Deletes the note asynchronous.
		/// </summary>
		/// <param name="noteId">The note identifier.</param>
		/// <param name="userName">The user id.</param>
		/// <returns>
		/// The boolean for success/failure
		/// </returns>
		/// <exception cref="System.ArgumentNullException">noteId</exception>
		public Task<bool> DeleteNoteAsync(int noteId, string userName)
		{
			HandleUserValidation(userName);
			if (noteId > 0)
			{
				return this._notesDataService.DeleteNoteAsync(noteId, userName);
			}

			throw new ArgumentNullException(nameof(noteId), ExceptionConstants.NoteNotFoundException);
		}

		/// <summary>
		/// Gets all notes asynchronous.
		/// </summary>
		/// <param name="userName">The user id.</param>
		/// <returns>
		/// The list of note entity
		/// </returns>
		public Task<IEnumerable<Note>> GetAllNotesAsync(string userName)
		{
			HandleUserValidation(userName);
			return this._notesDataService.GetAllNotesAsync(userName);
		}

		/// <summary>
		/// Gets the note asynchronous.
		/// </summary>
		/// <param name="noteId">The note id.</param>
		/// <param name="userName">The user id.</param>
		/// <returns>
		/// The note entity
		/// </returns>
		/// <exception cref="System.ArgumentNullException">noteId</exception>
		public Task<Note> GetNoteAsync(int noteId, string userName)
		{
			HandleUserValidation(userName);
			if (noteId > 0)
			{
				return this._notesDataService.GetNoteAsync(noteId, userName);
			}

			throw new ArgumentNullException(nameof(noteId), ExceptionConstants.NoteNotFoundException);
		}

		/// <summary>
		/// Updates the note asynchronous.
		/// </summary>
		/// <param name="updatedNote">The updated note.</param>
		/// <returns>
		/// The updated note
		/// </returns>
		/// <exception cref="System.ArgumentNullException">updatedNote</exception>
		public Task<Note> UpdateNoteAsync(UpdateNoteDTO updatedNote)
		{
			HandleUserValidation(updatedNote.UserName);
			if (updatedNote is not null)
			{
				return this._notesDataService.UpdateNoteAsync(updatedNote);
			}

			throw new ArgumentNullException(nameof(updatedNote), ExceptionConstants.NoteNotFoundException);
		}

		#region PRIVATE Methods

		/// <summary>
		/// Handles user validation.
		/// </summary>
		/// <param name="userName">The user name.</param>
		/// <exception cref="ArgumentNullException">ArgumentNullException error.</exception>
		private static void HandleUserValidation(string userName)
		{
			if (string.IsNullOrEmpty(userName))
			{
				throw new ArgumentNullException(nameof(userName), ExceptionConstants.AttemptedToEnterBlankUserMessageConstant);
			}
		}

		#endregion
	}
}

File: NotesTracker.Web/src/app/services/notes.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, first, firstValueFrom, map } from 'rxjs';

import { ApiUrls, ExceptionMessages } from '@shared/notestracker.constants';
import { Notes } from '@models/notes.model';
import { ResponseDTO } from '@models/dto/response-dto.class';
import { NoteDTO } from '@models/dto/note-dto.class';
import { UpdateNoteDTO } from '@models/dto/update-note-dto.class';
import { environment } from '@environments/environment';

/**
 * The Notes Service class.
 */
@Injectable({
  providedIn: 'root',
})
class NotesService {
  /**
   * The api url.
   */
  private apiUrl: string = `${environment.apiBaseUrl}/${ApiUrls.Notes.BaseRoute}`;

  /**
   * The notes routes.
   */
  private notesRoutes = ApiUrls.Notes;

  /**
   * Initializes a new instance of `NotesService`.
   * @param http The http client.
   */
  constructor(private http: HttpClient) {}

  /**
   * Gets all the notes data asynchronously.
   * @returns The api response
   */
  public async getAllNotesAsync(): Promise<Notes[]> {
    const getNotesUrl: string = `${this.apiUrl}${this.notesRoutes.GetAllNotes_ApiRoute}`;
    var response = await firstValueFrom(
      this.http.get<ResponseDTO>(getNotesUrl)
    );
    if (response?.isSuccess && response?.responseData !== null) {
      return response.responseData as Notes[];
    } else {
      throw new Error(response.responseData);
    }
  }

  /**
   * Gets the note data by id asynchronously.
   * @param noteId The note id.
   * @returns The api response.
   */
  public async getNoteByIdAsync(noteId: number): Promise<Notes> {
    const getNoteByIdUrl: string = `${this.apiUrl}${this.notesRoutes.GetNoteById_ApiRoute}?noteId=${noteId}`;
    const response = await firstValueFrom(
      this.http.get<ResponseDTO>(getNoteByIdUrl)
    );
    if (response?.isSuccess && response?.responseData !== null) {
      return response.responseData as Notes;
    } else {
      throw new Error(ExceptionMessages.NoteFetchFailedMessage);
    }
  }

  /**
   * Adds a new note asynchronously.
   * @param newNote The new note data.
   * @returns The api response.
   */
  public async addNewNoteAsync(newNote: NoteDTO): Promise<boolean> {
    const addNewNoteUrl: string = `${this.apiUrl}${this.notesRoutes.AddNewNote_ApiRoute}`;
    var response = await firstValueFrom(
      this.http.post<ResponseDTO>(addNewNoteUrl, newNote)
    );
    if (response?.isSuccess && response?.responseData !== null) {
      return response.responseData as boolean;
    } else {
      throw new Error(ExceptionMessages.AddingNoteFailedMessage);
    }
  }

  /**
   * Updates an existing note data asynchronously.
   * @param updatedNote The updated note data.
   * @returns The api response.
   */
  public async updateExistingNoteAsync(
    updatedNote: UpdateNoteDTO
  ): Promise<Notes> {
    const updateNoteUrl: string = `${this.apiUrl}${this.notesRoutes.UpdateNote_ApiRoute}`;
    const response = await firstValueFrom(
      this.http.post<ResponseDTO>(updateNoteUrl, updatedNote)
    );
    if (response?.isSuccess && response?.responseData) {
      return response?.responseData as Notes;
    } else {
      throw new Error(ExceptionMessages.NoteFetchFailedMessage);
    }
  }

  /**
   * Deletes an existing note by id asynchronously.
   * @param noteId The note id.
   * @returns The api response.
   */
  public async deleteNoteAsync(noteId: number): Promise<boolean> {
    const deleteNoteUrl: string = `${this.apiUrl}${this.notesRoutes.DeleteNote_ApiRoute}`;
    const response = await firstValueFrom(
      this.http.post<ResponseDTO>(deleteNoteUrl, noteId)
    );
    if (response?.isSuccess && response?.responseData !== null) {
      return response.responseData as boolean;
    } else {
      throw new Error(ExceptionMessages.NoteFetchFailedMessage);
    }
  }
}

export { NotesService };
